File: py/connect4/play_vs_cpu.py
#!/usr/bin/env python3
import argparse
import os
import sys
from typing import Optional

import numpy as np

sys.path.append(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
from game_runner import GameRunner
from connect4.perfect_player import PerfectPlayerParams, PerfectPlayer
from connect4.game_logic import Color, C4GameState
from connect4.human_tui_player import C4HumanTuiPlayer
from connect4.nnet_player import NNetPlayer, NNetPlayerParams


class Args:
    model_file: str = 'c4_model.pt'
    debug_filename: str = None
    verbose: bool = False
    perfect: bool = False
    my_starting_color: Optional[Color] = None
    neural_network_only: bool = False
    num_mcts_iters: int = 100
    temperature: float = 0.0

    @staticmethod
    def str_to_color(s: Optional[str]) -> Optional[Color]:
        if s is None:
            return None
        if s == 'R':
            return C4GameState.RED
        if s == 'Y':
            return C4GameState.YELLOW
        raise Exception(f'Invalid color -c {s}')

    @staticmethod
    def load(args):
        Args.model_file = args.model_file
        Args.debug_filename = args.debug_filename
        Args.verbose = args.verbose
        Args.perfect = args.perfect
        Args.my_starting_color = Args.str_to_color(args.my_starting_color)
        Args.neural_network_only = args.neural_network_only
        Args.num_mcts_iters = args.num_mcts_iters
        Args.temperature = args.temperature


def load_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("-m", "--model-file", default=Args.model_file,
                        help='model output location (default: %(default)s)')
    parser.add_argument("-d", "--debug-filename", help='debug output file')
    parser.add_argument("-v", "--verbose", action='store_true', help='verbose mode')
    parser.add_argument("-p", "--perfect", action='store_true', help='play against perfect player')
    parser.add_argument("-c", "--my-starting-color", help='my starting color (R or Y). Default: random')
    parser.add_argument("-o", "--neural-network-only", action='store_true', help='neural network only')
    parser.add_argument("-n", "--num-mcts-iters", default=Args.num_mcts_iters, type=int,
                        help='num mcts iterations to do per move (default: %(default)s)')
    parser.add_argument("-t", "--temperature", default=Args.temperature, type=float,
                        help='temperature. Must be >=0. Higher=more random play (default: %(default)2f)')

    args = parser.parse_args()
    Args.load(args)


def main():
    load_args()

    human = C4HumanTuiPlayer()

    if Args.perfect:
        cpu = PerfectPlayer(PerfectPlayerParams())
    else:
        params = NNetPlayerParams(
            model_file=Args.model_file,
            debug_filename=Args.debug_filename,
            verbose=Args.verbose,
            neural_network_only=Args.neural_network_only,
            num_mcts_iters=Args.num_mcts_iters,
            temperature=Args.temperature)
        cpu = NNetPlayer(params)

    my_color = np.random.randint(0, 2) if Args.my_starting_color is None else Args.my_starting_color
    cpu_color = 1 - my_color
    players = [None, None]
    players[my_color] = human
    players[cpu_color] = cpu

    runner = GameRunner(C4GameState, players)
    result = runner.run()

    if result[my_color] == 1:
        print('Congratulations! You win!')
    elif result[cpu_color] == 1:
        print('Sorry! You lose!')
    else:
        print('The game has ended in a draw!')


if __name__ == '__main__':
    main()

File: cpp/connect4/main/PlayVsCpu.cpp
#include <iostream>

#include <boost/program_options.hpp>

#include <common/BasicTypes.hpp>
#include <common/GameRunner.hpp>
#include <common/HumanTuiPlayer.hpp>
#include <common/MctsPlayer.hpp>
#include <connect4/C4CheatingHumanTuiPlayer.hpp>
#include <connect4/C4Constants.hpp>
#include <connect4/C4GameState.hpp>
#include <connect4/C4PerfectPlayer.hpp>
#include <connect4/C4Tensorizor.hpp>
#include <util/Config.hpp>
#include <util/Exception.hpp>
#include <util/Random.hpp>

namespace po = boost::program_options;
namespace po2 = boost_util::program_options;

struct Args {
  std::string my_starting_color;
  bool perfect;
  bool cheat_mode;

  auto make_options_description() {
    po2::options_description desc("PlayVsCpu options");

    return desc
        .template add_option<"my-starting-color", 's'>(po::value<std::string>(&my_starting_color),
            "human's starting color (R or Y). Default: random")
        .template add_option<"perfect", 'p'>(po::bool_switch(&perfect)->default_value(false),
            "play against perfect player")
        .template add_option<"cheat-mode", 'C'>(po::bool_switch(&cheat_mode)->default_value(false),
            "show winning moves")
        ;
  }
};

using GameState = c4::GameState;
using Tensorizor = c4::Tensorizor;
using Mcts = common::Mcts<GameState, Tensorizor>;
using MctsPlayer = common::MctsPlayer<GameState, Tensorizor>;

common::player_index_t parse_color(const std::string& str) {
  if (str == "R") return c4::kRed;
  if (str == "Y") return c4::kYellow;
  if (str.empty()) return util::Random::uniform_sample(0, c4::kNumPlayers);
  throw util::Exception("Invalid --my-starting-color/-s value: \"%s\"", str.c_str());
}

int main(int ac, char* av[]) {
  Mcts::Params mcts_params(Mcts::kCompetitive);
  MctsPlayer::Params mcts_player_params(MctsPlayer::kCompetitive);
  c4::PerfectPlayParams perfect_play_params;
  Args args;

  po2::options_description raw_desc("General options");
  auto desc = raw_desc.template add_option<"help", 'h'>("help")
      .add(mcts_params.make_options_description())
      .add(mcts_player_params.make_options_description())
      .add(perfect_play_params.make_options_description())
      .add(args.make_options_description());

  po::variables_map vm;
  po::store(po::command_line_parser(ac, av).options(desc).run(), vm);
  po::notify(vm);

  if (vm.count("help")) {
    std::cout << desc << std::endl;
    return 0;
  }

  c4::Player* human;
  if (args.cheat_mode) {
    mcts_player_params.verbose = true;
    human = new c4::CheatingHumanTuiPlayer(perfect_play_params);
  } else {
    using C4HumanTuiPlayer = common::HumanTuiPlayer<c4::GameState>;
    human = new C4HumanTuiPlayer();
  }

  c4::Player* cpu;
  if (args.perfect) {
    cpu = new c4::PerfectPlayer(perfect_play_params);
  } else {
    cpu = new MctsPlayer(mcts_player_params, mcts_params);
  }

  common::player_index_t my_color = parse_color(args.my_starting_color);
  common::player_index_t cpu_color = 1 - my_color;

  using player_t = common::AbstractPlayer<c4::GameState>;
  using player_array_t = std::array<player_t*, c4::kNumPlayers>;

  player_array_t players;
  players[my_color] = human;
  players[cpu_color] = cpu;

  using GameRunner = common::GameRunner<c4::GameState>;
  GameRunner runner(players);
  auto outcome = runner.run(GameRunner::kFixedPlayerSeats);

  if (outcome[my_color] == 1) {
    std::cout << "Congratulations, you win!" << std::endl;
  } else if (outcome[cpu_color] == 1) {
    std::cout << "Sorry! You lose!" << std::endl;
  } else {
    std::cout << "The game has ended in a draw!" << std::endl;
  }

  delete cpu;
  delete human;

  return 0;
}
